web-runtime/src/module/wit_instance.rs

```diff
@@ -0,0 +1,165 @@
+#[derive(Clone)]
+pub(super) struct ITInstance {
+    /// IT functions indexed by id.
+    funcs: HashMap<usize, WITFunction>,
+
+    /// IT memories.
+    memories: Vec<WITMemory>,
+
+    /// All record types that instance contains.
+    record_types_by_id: MRecordTypes,
+}
+
+impl ITInstance {
+    pub(super) fn new(wasmer_instance: &WasmerInstance, wit: &MITInterfaces<'_>) -> MResult<Self> {
+        let exports = Self::extract_raw_exports(wasmer_instance, wit)?;
+        let memories = Self::extract_memories(wasmer_instance);
+        js_log(&format!("executed Self::extract_memories"));
+
+        let funcs = exports;
+
+        let record_types_by_id = Self::extract_record_types(wit);
+        js_log(&format!("executed Self::extract_record_types"));
+
+        Ok(Self {
+            funcs,
+            memories,
+            record_types_by_id,
+        })
+    }
+
+    fn extract_raw_exports(
+        wasmer_instance: &WasmerInstance,
+        it: &MITInterfaces<'_>,
+    ) -> MResult<HashMap<usize, WITFunction>> {
+        let module_exports = &wasmer_instance.exports;
+
+        it.exports()
+            .enumerate()
+            .map(|(export_id, export)| {
+                let export_func = module_exports.get(export.name)?;
+                unsafe {
+                    // TODO: refactor this with new Wasmer API when it is ready
+                    // here it is safe because dyn func is never lives WITInstance
+                    let export_func =
+                        std::mem::transmute::<DynFunc<'_>, DynFunc<'static>>(export_func);
+                    Ok((
+                        export_id,
+                        WITFunction::from_export(export_func, export.name.to_string())?,
+                    ))
+                }
+            })
+            .collect()
+    }
+
+    fn extract_memories(wasmer_instance: &WasmerInstance) -> Vec<WITMemory> {
+        use crate::marine_js::Export::Memory;
+
+        let memories = wasmer_instance
+            .exports()
+            .filter_map(|(_, export)| match export {
+                Memory => Some(WITMemory::new(wasmer_instance.module_name.clone())),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
+        memories
+    }
+
+    fn extract_record_types(wit: &MITInterfaces<'_>) -> MRecordTypes {
+        let (record_types_by_id, _) = wit.types().fold(
+            (HashMap::new(), 0u64),
+            |(mut record_types_by_id, id), ty| {
+                match ty {
+                    ITAstType::Record(record_type) => {
+                        record_types_by_id.insert(id, record_type.clone());
+                    }
+                    ITAstType::Function { .. } => {}
+                };
+                (record_types_by_id, id + 1)
+            },
+        );
+
+        record_types_by_id
+    }
+}
+
+impl wasm::structures::Instance<ITExport, WITFunction, WITMemory, WITMemoryView> for ITInstance {
```

Comment:
all logs here look excess for master

Reply:
Sure, i will remove any debug output i find